crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2615,36 +2615,112 @@ impl<'db> Type<'db> {
+                    match instance_ty.try_call_dunder(db, "__init__", arguments) {
+                        Err(CallDunderError::Call(mut error)) => {
+                            // If the call to `__init__` fails we want to return inner error directly
+                            // but ensure that the fallback return type is the instance type.
+
+                            match error {
+                                CallError::NotCallable { .. } => {
+                                    // Instead of trying to emit a diagnostic at call site, these should be
+                                    // caught at the definition site of the class. Here we return Ok with the
+                                    // instance type to ensure that the return type is correct.
+                                    Ok(CallOutcome::Single(CallBinding::from_return_type(
+                                        instance_ty,
+                                    )))
+                                }
+                                CallError::PossiblyUnboundDunderCall { .. } => {
+                                    tracing::warn!("Inferred `__init__` as possibly unbound on a class. All classes have `object` in MRO, so this is not possible. This indicates an error in typeshed.");
+
+                                    // We return None (see the logic for `NotCallable` above).
+                                    Ok(CallOutcome::Single(CallBinding::from_return_type(
+                                        instance_ty,
+                                    )))
+                                }
+                                CallError::Union(_) | CallError::BindingError { .. } => {
+                                    // Replace the return type and return as Err
+                                    error.set_return_type(instance_ty);
+                                    Err(error)
+                                }
+                            }
+                        }
+                        // If we are using vendored typeshed, it should be impossible to have missing
+                        // or unbound `__init__` methods on classes, as all classes have `object` in MRO,
+                        // hence we emit user warnings for these cases.
+                        Err(CallDunderError::PossiblyUnbound(mut outcome)) => {
+                            tracing::warn!("Inferred `__init__` as possibly unbound on a class. All classes have `object` in MRO, so this is not possible. This indicates an error in typeshed.");
+
+                            outcome.set_return_type(instance_ty);
+                            Ok(outcome)
+                        }
+                        Err(CallDunderError::MethodNotAvailable) => {
+                            tracing::warn!("Inferred `__init__` as unbound on a class. All classes have `object` in MRO, so this is not possible. This indicates an error in typeshed.");
```

Comment:
I have a mild preference for not even emitting these warnings. Having a custom typeshed without `object.__init__` is really a harmless situation that we can easily fall back from, so why create extra noise?